### PR TITLE
[DRAFT] Updates to Core and WebAPI integrations to support SE

### PR DESCRIPTION
**Pull request by DDEfromOR**
This PR is to preview the changes to bring SE support to Bot Builder for .NetFx and .NetCore.

The base branch this PR is against is based off of 4.3.

The changes in this branch bring support for using SE as a communication protocol, and support for all of the Bot Connector calls to the channel.

### CODE REVIEW

File: libraries/integration/Microsoft.Bot.Builder.Integration.AspNet.Core/StreamingExtensions/BotFrameworkStreamingExtensionsAdapter.cs [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+            var request = Request.CreateGet(route);
+
+            return await SendRequestAsync<IList<ChannelAccount>>(request).ConfigureAwait(false);
+        }
+
+        private async Task<T> SendRequestAsync<T>(Request request)
+        {
+            try
+            {
+                var serverResponse = await _server.SendAsync(request).ConfigureAwait(false);
```
Comment (Jeffders) [inline libraries/integration/Microsoft.Bot.Builder.Integration.AspNet.Core/StreamingExtensions/BotFrameworkStreamingExtensionsAdapter.cs]: I'm adding a cancellationToken parameter here...you'll want to pass one in, and then catch "TaskCancelledException" and deal with timeouts

File: libraries/integration/Microsoft.Bot.Builder.Integration.AspNet.Core/StreamingExtensions/StreamingExtensionRequestHandler.cs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Bot.Protocol;
+using Microsoft.Bot.Protocol.WebSockets;
+using Microsoft.Bot.Schema;
+using Newtonsoft.Json;
+
+namespace Microsoft.Bot.Builder.Integration.AspNet.Core.StreamingExtensions
+{
+    internal class StreamingExtensionRequestHandler : RequestHandler
```
Comment (cleemullins) [inline libraries/integration/Microsoft.Bot.Builder.Integration.AspNet.Core/StreamingExtensions/StreamingExtensionRequestHandler.cs]: I'm not understanding the role this class is intended to play. Could you add signifigant comments explaining what this class does, and how it fits into the larger picture?

File: libraries/integration/Microsoft.Bot.Builder.Integration.AspNet.WebApi/Microsoft.Bot.Builder.Integration.AspNet.WebApi.csproj [hunk 3]
```diff
@@ -43,7 +43,9 @@
 		<PackageReference Include="AsyncUsageAnalyzers" Version="1.0.0-alpha003" PrivateAssets="all" />
 		<PackageReference Include="Microsoft.AspNet.WebApi" Version="5.2.6" />
     <PackageReference Include="Microsoft.Bot.Builder" Condition=" '$(PackageVersion)' == '' " Version="4.0.0-local" />
-    <PackageReference Include="Microsoft.Bot.Builder" Condition=" '$(PackageVersion)' != '' " Version="$(PackageVersion)" />
+    <PackageReference Include="Microsoft.Bot.Protocol" Version="1.0.0" />
```
Comment (cleemullins) [inline libraries/integration/Microsoft.Bot.Builder.Integration.AspNet.WebApi/Microsoft.Bot.Builder.Integration.AspNet.WebApi.csproj]: You're making some key changes to our project files, and I'm guessing (from the diff) you're not quite sure why the deleted lines are there in the first place. Come see me, and I'll explain. :)

File: libraries/integration/Microsoft.Bot.Builder.Integration.AspNet.WebApi/StreamingExtensions/BotFrameworkStreamingExtensionsAdapter.cs [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+using System.Threading.Tasks;
+using Microsoft.Bot.Protocol;
+using Microsoft.Bot.Protocol.WebSockets;
+using Microsoft.Bot.Schema;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+using Microsoft.Rest;
+using Newtonsoft.Json;
+
+namespace Microsoft.Bot.Builder.Integration.AspNet.WebApi.StreamingExtensions
```
Comment (cleemullins) [inline libraries/integration/Microsoft.Bot.Builder.Integration.AspNet.WebApi/StreamingExtensions/BotFrameworkStreamingExtensionsAdapter.cs]: Before delving too far into WebAPI, might I suggest we get to closure on the .Net Core implementation first? I suspect nearly all of my questions/comments there apply here as well.

File: libraries/integration/Microsoft.Bot.Builder.Integration.AspNet.WebApi/StreamingExtensions/IBotFrameworkStreamingExtensionsAdapter.cs [hunk 5]
```diff
@@ -0,0 +1,339 @@
+﻿using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Bot.Schema;
+
+namespace Microsoft.Bot.Builder.Integration.AspNet.WebApi.StreamingExtensions
+{
+    public interface IBotFrameworkStreamingExtensionsAdapter
```
Comment (cleemullins) [inline libraries/integration/Microsoft.Bot.Builder.Integration.AspNet.WebApi/StreamingExtensions/IBotFrameworkStreamingExtensionsAdapter.cs]: I don't see how this class is different from the .Net Core implementation. Could you explain?

File: libraries/integration/Microsoft.Bot.Builder.Integration.AspNet.Core/StreamingExtensions/BotFrameworkWebSocketAdapter.cs [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+            {
+                httpRequest.HttpContext.Response.StatusCode = (int)HttpStatusCode.BadRequest;
+                await httpRequest.HttpContext.Response.WriteAsync("Upgrade to WebSocket required.").ConfigureAwait(false);
+                return;
+            }
+
+            await CreateWebSocketConnectionAsync(httpRequest.HttpContext, bot).ConfigureAwait(false);
+        }
+
+        public async Task CreateWebSocketConnectionAsync(HttpContext httpContext, IBot bot)
```
[the remark this answers is not in the document]
Comment (DDEfromOR) [reply libraries/integration/Microsoft.Bot.Builder.Integration.AspNet.Core/StreamingExtensions/BotFrameworkWebSocketAdapter.cs]: True, and fixed.

File: libraries/integration/Microsoft.Bot.Builder.Integration.AspNet.Core/StreamingExtensions/BotFrameworkWebSocketAdapter.cs [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+        private readonly IChannelProvider _channelProvider;
+        private readonly ICredentialProvider _credentialProvider;
+
+        public BotFrameworkWebSocketAdapter(ICredentialProvider credentialProvider, IChannelProvider channelProvider = null, ILogger<BotFrameworkWebSocketAdapter> logger = null)
+        {
+            this._credentialProvider = credentialProvider;
+            this._channelProvider = channelProvider;
+        }
+
+        public ConcurrentDictionary<string, WebSocketServer> Connections { get; set; }
```
[the remark this answers is not in the document]
Comment (DDEfromOR) [reply libraries/integration/Microsoft.Bot.Builder.Integration.AspNet.Core/StreamingExtensions/BotFrameworkWebSocketAdapter.cs]: Cruft left over from an earlier design iteration, you're right that it's no longer used and doesn't need to exist. Removed!

File: libraries/Microsoft.Bot.Builder/Microsoft.Bot.Builder.csproj [hunk 8]
```diff
@@ -39,4 +39,13 @@
 		<ProjectReference Include="..\Microsoft.Bot.Connector\Microsoft.Bot.Connector.csproj" />
 	</ItemGroup>
 
+	<ItemGroup>
+	  <Reference Include="Microsoft.Bot.Protocol">
+	    <HintPath>..\..\..\Intercom\Microsoft.DirectLineFlex\Microsoft.Bot.Protocol\bin\Debug\netcoreapp2.2\Microsoft.Bot.Protocol.dll</HintPath>
```
[the remark this answers is not in the document]
Comment (DDEfromOR) [reply libraries/Microsoft.Bot.Builder/Microsoft.Bot.Builder.csproj]: Yikes, there shouldn't be any changes to Microsoft.Bot.Builder in here, reverting.

**Pull request closed without merging**